rfcs/convergence/internationalization.md

```diff
@@ -0,0 +1,217 @@
+
+```js
+import { defaultLocale } from './defaultStrings.ts';
+
+export const ProviderContext =
+  React.createContext <
+  ProviderContextValue >
+  {
+    targetDocument: typeof document === 'object' ? document : undefined,
+    dir: 'ltr',
+    locale: defaultLocale,
+  };
+```
+
+Then, each component would take a `strings` prop where strings could be directly passed in, and also provide component-specific defaults when needed. The `strings` prop would also be used for the author to pass in strings that don't have a default, e.g. error messaging and instance-specific labels.
+
+react-dropdown/src/localeStrings.ts (specific strings are just for this example)
+
+```js
+export const dropdownDefaultStrings: Partial<DropdownStrings> {
+  loadMore: 'Load more options',
+  invalidEntryError: 'Your search did not match any options'
+}
+```
+
+react-dropdown/src/components/useDropdown.tsx (very simplified)
+
+```js
+import { dropdownDefaultStrings } from '../localeStrings';
+export const useDropdown = props => {
+  const { locale } = useFluent();
+  const dropdownStrings = {
+    ...locale.strings.global, // only if this component actually uses global strings
+    ...locale.strings.Dropdown,
+    ...props.strings,
+  };
+
+  state.components.arrow['aria-label'] = dropdownStrings.expandCollapse;
+};
+```
+
+## Concrete actions for vNext beta
+
+The only action needed for beta release should be to ensure all strings provided through component props (excluding children/child content) are defined in a single `strings` property.
+
+For example, the `PresenceBadge` (and also components like `Avatar` that use it) would need something like this for the strings prop (likely with defaults in a component-specific file for `PresenceBadge`):
+
+```js
+interface PresenceBadgeStrings {
+  statusBusy: string;
+  statusOutOfOffice: string;
+  statusAway: string;
+  statusAvailable: string;
+  statusOffline: string;
+  statusDnD: string;
+}
+
+interface PresenceBadgeProps {
+  // ...etc
+
+  strings?: PresenceBadgeStrings;
+}
+
+interface AvatarStrings extends PresenceBadgeStrings {
+  active: string;
+  inactive: string;
+}
+
+// defined in a default strings file
+const presenceDefaultStrings: PresenceBadgeStrings = {
+  statusBusy: 'busy',
+  statusOutOfOffice: 'out of office',
+  statusAway: 'away',
+  statusAvailable: 'available',
+  statusOffline: 'offline',
+  statusDnD: 'do not disturb',
+};
+```
+
+## Future possibilities
+
+Another possibility is that we could also provide a way to (optionally) pass a function (e.g. `localizeString`) through `ProviderContext` to better support libraries like `react-i18next` and `react-intl`. Then, if that function is provided, we call it with any given string as an argument. That would mean someone using `react-i18next` could do something like this:
+
+App.ts:
+
+```js
+import { useTranslation } from 'react-i18next';
+
+const App = () => {
+  const { t, i18n } = useTranslation('es-ES');
```

Comment:
we would force users to use our localization keys for their applications or force them to build a mapping layer between our strings and theirs if we support i18n this way. Is this acceptable for our partners ?